---
title: "Practical 1: Introduction"
subtitle: "EMBL Course: Data mining and integration with networks"
author: "Aaron Brooks"
date: "This should take about 25 minutes"
output: 
  html_document:
    css: include/hint.css
    includes:
      in_header: include/hint.html
    toc: true
    toc_depth: 2
---

```{r, echo = FALSE, message = FALSE}
devtools::source_url("https://raw.githubusercontent.com/scalefreegan/Teaching/master/DataIntegration/scripts/readData.R")
```
[Companion lecture](http://scalefreegan.github.io/Teaching/DataIntegration/lectures/l1.html)

# Introduction to Clustering

Welcome to the first practical lesson for the **clustering** section of the EMBL course: *Data mining and integration with networks*. 

This document will introduce basic concepts related to clustering and provide a gentle introduction to clustering with a simple example. At the end of the practical we will read and process Jean-Karim's kernels so that they are ready to analyze in the subsequent tutorials.

# Format of the practicals

First, a word about the format of these practicals. 

The practicals are designed to allow self-guided exploration of the course materials. Each practical is paired with information from the preceeding lecture. You should be able to explore this document at your own pace and even refer to it again in the future. 

In addition to the text, be on the lookout for three special formats in these documents. These sections include scripts, excercises and solutions, respectively.   

**Scripts**

Scripts are denoted by blocked text. You can copy and paste these commands into the R terminal. Give the command below a try. 

```{r, eval=FALSE}
cat("Welcome to the course on clustering\n")
```

**Excercises**

Throughout the practical you will be asked to evaluate your understanding by writing code. You should solve these excercise by writing your own script to perform the requested operations or answer the question. Excercises are denoted by blocked quotation, as below.

> This is an excercise.
>
> You should write your own code to solve it.

**Solutions**

If you get stuck and/or to see the answer I generated, use the **solution** pulldown to reveal the answer to the problem, including accompanying code.

The solution pulldowns look like this. Give the one below a try:

<div id='hint'>
<h5 class="ui-closed" style="color:blue">Solution</h5>
####"Do. Or do not. There is no try."

![](http://a.dilcdn.com/bl/wp-content/uploads/sites/6/2013/11/01-400x225.jpg)
</div>

*Don't cheat (too much). These excercises are designed to make sure that you're keeping up. Give it a solid effort before revealing the answer. Yoda remembers...*

# Requirements

This tutorial requires a number of R packages. To install theses packages you can run the script included below.

*All packages should be installed on the computers used in the Data Integration course.*

```{r, eval=FALSE}
install.packages("devtools")
devtools::source_url("https://raw.githubusercontent.com/scalefreegan/Teaching/master/DataIntegration/scripts/installPackages.R")
```
    
# Clustering

## Why cluster?

## Clustering simple data

At it simplest

Some examples of 

```{r, eval=TRUE}
library(kernlab)
data(spirals)
```

## More complicated data

What if the pattern in the data is more complicated? 

Consider the following data set. How many clusters do you see?

```{r, eval=TRUE, echo=FALSE}
plot(spirals, pch=19, xlab="", ylab="")
```

> Plot the data `spirals`. You can load this data set with `data(spirals)`.
>
> Run the k-means algorithm (# clusters = 2) on the spiral data. Plot data with clusters annotated. What do you observe?

<div id='hint'>
<h5 class="ui-closed" style="color:blue">Solution</h5>
```{r, eval=TRUE}
km <- kmeans(spirals, centers = 2)

plot(spirals, col=km$cluster, pch=19)
```
</div>

> Run spectral clustering (# clusters = 2) on the spiral data. The spectral clustering algorithm is called with the function `specc`. Once again, plot the data with clusters annotated. 
>
> What do you observe now?

<div id='hint'>
<h5 class="ui-closed" style="color:blue">Solution</h5>
```{r, eval=TRUE}
sc <- specc(spirals, centers=2)

plot(spirals, col=sc, pch=19)
```
</div>


# Kernel data set

Finally, let's process Jean-Karim's similarity kernels so that they are ready for subsequent lectures. 

To remind you, Jean-Karim has provided us with several graph kernels that measure similarities between 120 human genes according to information derived from 5 sources. These sources include:

Type   | Description
-------|------------------------------------------------------------------------------------
BP     | Pairwise gene semantic similarities (Resnik score) in the biological process domain
CODA80 | Predicted protein interactions from domain co-occurrence/fusion events
HIPPO  | Protein interactions transferred from model organisms to human by orthology
PI     | Protein interactions in human (from multiple databases)
TM     | Protein interactions derived from the iHOP text mining method

The 120 genes in our data set were chosen for their role in the following biological processes:

- chromosome condensation
- DNA mismatch repair ([KEGG:hsa03430](http://www.genome.jp/dbget-bin/www_bget?pathway+hsa03430))
- mitotic spindle checkpoint
- endocytosis ([GO:0006897](https://www.ebi.ac.uk/QuickGO/GTerm?id=GO:0006897))
- signalling by BMP

The remaining `r length(unique(c(data$gene1,data$gene2)))-120` genes have a predicted interaction with one of the 120 genes according to one of the 5 measures described above.

The first thing we need to do is combine each of these kernels into a single matrix (graph). It is on this combined graph that we will perform clustering. 

## Get data

## Combine and normalize kernels

## Make a graph

```{r, message=FALSE}
library(igraph)
```


